couchdb/server/views.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+#
+import copy
+import logging
+
+from couchdb import json
+from couchdb.server.exceptions import QueryServerException, Error
+
+__all__ = ('map_doc',)
+
+log = logging.getLogger(__name__)
+
+
+def map_doc(server, doc):
+    """Applies available map functions to document.
+
+    :command: map_doc
+
+    :param server: Query server instance.
+    :type server: :class:`~couchdb.server.BaseQueryServer`
+
+    :param doc: Document object.
+    :type doc: dict
+
+    :return: List of key-value results for each applied map function.
+
+    :raises:
+        - :exc:`~couchdb.server.exceptions.Error`
+          If any Python exception occurs due mapping.
+    """
+    docid = doc.get('_id')
+    log.debug('Apply map functions to document `%s`:\n%s', docid, doc)
+    orig_doc = copy.deepcopy(doc)
+    map_results = []
+    _append = map_results.append
+    try:
+        for idx, func in enumerate(server.state['functions']):
+            # TODO: https://issues.apache.org/jira/browse/COUCHDB-729
+            # Apply copy.deepcopy for `key` and `value` to fix this issue
+            _append([[key, value] for key, value in func(doc) or []])
```

Comment:
There are some test case using normal function. Seems we need to support normal function?
e.g.
```python
    def test_map_doc_by_many_functions(self):
        """should apply multiple map functions to single document"""
        state.add_fun(
            self.server,
            'def mapfun(doc):\n'
            '  yield doc["_id"], "foo"\n'
            '  yield doc["_id"], "bar"'
        )
        state.add_fun(
            self.server,
            'def mapfun(doc):\n'
            '  yield doc["_id"], "baz"'
        )
        state.add_fun(
            self.server,
            'def mapfun(doc):\n'
            '  return [[doc["_id"], "boo"]]'
        )
        result = views.map_doc(self.server, {'_id': 'foo'})
        self.assertEqual(result, [[['foo', 'foo'], ['foo', 'bar']],
                                  [['foo', 'baz']], [['foo', 'boo']]])
```